### ISSUE TITLE
Invalid non-undefined typings with && any

### ISSUE THREAD

**Issue by deregtd** [Microsoft/TypeScript#20318]
**TypeScript Version:** 2.7.0-dev.20171128

**Code**

```ts
        let x: string|undefined;
        let y: any;
        let z: number = x && y;
```

**Expected behavior:**
Compile Error: undefined is not assignable to number.

**Actual behavior:**
AOK.

Presumably x && y resolves to 'undefined|any', i.e. 'any', and any can be assigned to number.  However, we found a few instances of this causing bugs in our codebase when devs had done the equivalent of:

```ts
let something: SomeTypeNonUndefinable = val && JSON.parse(val);
```

... which happily compiled with strict null checks and didn't complain, but allowed undefineds through into 'something'.

**Comment by mhegazy** [comment]
`JSON.parse` can return you `undefined` (e.g. if `val` was a function for instance). so the issue here is `any` and not that ` undefined | any` flattens to `any`. 

Maybe the safe thing to do is to have JSON.parse return something less permissive like `{}`. or `unknown` (https://github.com/Microsoft/TypeScript/issues/10715).

**Comment by deregtd** [comment]
You can't do that, though, unfortunately -- JSON.parse can parse anything:

```ts
typeof JSON.parse('4')
"number"
```

**Comment by deregtd** [comment]
Maybe it returns undefined|Object?

**Comment by mhegazy** [comment]
`Object` is not the right type.. you want the top type, `{} | undefined | null`

**Comment by mhegazy** [comment]
I do not think we can change `JSON.parse` at this point, or that would be a big breaking change. One option is for you to add an overload for JSON.parse that returns `{} | undefined | null` in your project.
